src/Api/PubnubApi/EndPoint/PubSub/SubscribeEndpoint.cs

```diff
@@ -0,0 +1,219 @@
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+using System.Net;
+using System.Globalization;
+using PubnubApi.EventEngine.Subscribe;
+using PubnubApi;
+using PubnubApi.EventEngine.Core;
+using PubnubApi.EventEngine.Subscribe.Events;
+using PubnubApi.EventEngine.Subscribe.States;
+using PubnubApi.EventEngine.Subscribe.Common;
+
+namespace PubnubApi.EndPoint
+{
+    public class SubscribeEndpoint<T>: ISubscribeOperation<T>
+    {
+        private readonly PNConfiguration config;
+        private readonly IJsonPluggableLibrary jsonLibrary;
+        private readonly IPubnubUnitTest unit;
+        private readonly IPubnubLog pubnubLog;
+        private readonly EndPoint.TelemetryManager pubnubTelemetryMgr;
+        private readonly EndPoint.TokenManager pubnubTokenMgr;
+
+        private List<string> subscribeChannelNames = new List<string>();
+        private List<string> subscribeChannelGroupNames = new List<string>();
```

Comment:
I don't know if is it a good practice or not but I've seen in our code many occurrences of `IEnumerable` instead of using lists directly. Should we continue this trend?

Reply:
Here the purpose of using `List<string>` is, We are going to modify the channels/groups (based on presence channels/groups subscribed or not)
Having `List` type here makes add/remove or indexer available for further operations.

Not sure about other places, But If we declare `IEnumerable` here  then we will need to convert it to list/array whenever we need to access any element /update it

Reply:
So It's based on usage we should assign IEnumerable / List/ Array.